02_knight_and_princess/anton_potapov/graph_create.cpp

```diff
@@ -0,0 +1,115 @@
+#include <cassert>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "properties.hpp"
+
+using VertexId = int;
+using EdgeId = int;
+
+struct Vertex {
+  VertexId id;
+  Camp camp_info;
+  Vertex(int vertex_id, Camp c_info = Camp())
+      : id(vertex_id), camp_info(c_info) {}
+};
+
+struct Edge {
+  EdgeId id;
+  VertexId vertex1_id, vertex2_id;
+  Boss boss_info;
+  Edge(EdgeId e_id, VertexId ver1, VertexId ver2, Boss b_info = Boss())
+      : id(e_id), vertex1_id(ver1), vertex2_id(ver2), boss_info(b_info) {}
+};
+
+class Graph {
+ public:
+  Graph() = default;
+  Graph(const std::vector<Vertex>& vertices, const std::vector<Edge>& edges)
```

Comment:
Усложняешь себе жизнь вот этим конструктором. Ну раз уж хочешь... В таком случае, нужно бы проверить на валидность входных данных. Все ли вершины связаны гранями? Я бы на твоем месте, в данном случае упростил бы, и передавал просто список граней, из которых уже заполнил `adjacency_list_`. Зачем тебе вообще передавать сюда вершины?
```suggestion
  Graph(const std::vector<Edge>& edges)
```

Reply:
Вот этот момент не очень понял. В таком случае мы не сможем сконструировать граф с изолированнными вершинами, что в итоге нам даст неочевидное ограничение на хранимый граф и это придется где-то специфицировать. На связность пришлось бы проверять в любом случае. 
К тому же у меня в векторе `edges`, как в векторе структур `Edge`, не хранятся сами вершины как структуры `Vertex`, а только как их id. То есть если бы мы хотели построить граф со сложной информацией в вершинах, нам бы это удалось только после того, как мы сконструировали этот граф по ребрам.
Добавил проверку на соответствие вектора ребер вектору вершин

Reply:
Вообще, по хорошему, `id` - это внутренняя структура графа, которая не должна иметь выход наружу. `id` должен генерироваться внутри графа, а интерфейс графа должен давать пользователю найти ту вершину, которая ему нужна.
Чтобы лучше понять, о чем я говорю, можешь посмотреть на ПР Новикова - #50

Reply:
То есть чтобы пользователь не смог все сломать, передав графу две вершины с одинаковым `id`? Makes sense, наверное нужно сделать тоже самое и для `EdgeId`

Reply:
Аа, можно при добавлении вершины возвращать ее `VertexId`. Now it's all makes sense again

Reply:
Хотя лучше не `VertexId`, а какую-нибудь структуру типо `VertexProperties`, по которой потом можно будет найти эту вершину